core/src/main/scala/kafka/log/LogCleanerManager.scala

```diff
@@ -537,6 +537,13 @@ private[log] object LogCleanerManager extends Logging {
         warn(s"The last checkpoint dirty offset for partition $topicPartition is $checkpointDirtyOffset, " +
           s"which is larger than the log end offset ${log.logEndOffset}. Resetting to the log start offset $logStartOffset.")
         logStartOffset
+      } else if (checkpointDirtyOffset > log.activeSegment.baseOffset && logStartOffset > log.activeSegment.baseOffset) {
```

Comment:
I'm not sure this check makes sense. If the log start offset has gotten ahead of the active segment base offset, the odds are that it has gotten ahead of the checkpoint offset as well, which means we will hit the first check above. 

I think really what we want from this function is to retain a valid offset range. Valid in this case means 1) that the upper limit is greater than or equal to the lower limit and 2) that the range is a valid portion of the log. The existing checks here guarantee that the lower limit is greater than or equal to the log start offset, but the upper limit may violate both 1 and 2 when the log start offset has been incremented into the active segment.

I would suggest a simpler fix, which is just to force the upper limit into a valid range. In other words, we return the following instead of the current result:
```scala
    (firstDirtyOffset, math.max(firstDirtyOffset, firstUncleanableDirtyOffset))
```
We are then guaranteed that `logStartOffset <= firstDirtyOffset <= firstUncleanableDirtyOffset <= activeSegment.baseOffset <= logEndOffset`.